documentation/projects/proposals/search_relevancy_sandbox/20230518-implementation_plan_staging_index_rapid_iteration.md

```diff
@@ -0,0 +1,380 @@
+   and
+   [`create_and_populate_filtered_index` on the ingestion server](https://github.com/WordPress/openverse/blob/3fcce5ade2165955db5bbcb4f679257b3260547b/ingestion_server/ingestion_server/indexer.py#L529)
+   for an example of how this is used.
+
+### DAG
+
+Once the parameters are provided, the DAG will execute the following steps:
+
+1. (For `production` only) Check that there are no other reindexing jobs
+   currently underway. This can be done by checking whether the
+   `{media_type}_data_refresh` and `create_filtered_{media_type}_index` DAGs are
+   currently running. If either of them are, this DAG should fail immediately
+   with an appropriate error message.
+2. Create the index configuration. If `override_config` is supplied, the
+   `index_config` parameter will be used as the entire configuration. If not,
+   the `index_config` parameter will be merged with the existing index settings.
+   See [Merging policy](#merging-policy) for how this merge will be performed.
+   Below are the steps for gathering the current index settings:
+   1. Get the current index information. This will be done using the
+      [`ElasticsearchPythonHook`](https://airflow.apache.org/docs/apache-airflow-providers-elasticsearch/stable/_api/airflow/providers/elasticsearch/hooks/elasticsearch/index.html#airflow.providers.elasticsearch.hooks.elasticsearch.ElasticsearchPythonHook),
+      specifically the
+      [`index.get` function](https://elasticsearch-py.readthedocs.io/en/v8.8.0/api.html#elasticsearch.client.IndicesClient.get).
+   2. Extract the relevant settings from the response. They come back in a form
+      that differs from what is rendered by the
+      [`es_mapping.py`](https://github.com/WordPress/openverse/blob/main/ingestion_server/ingestion_server/es_mapping.py)
+      file which generates the index settings on the ingestion server.
+      `es_mapping::index_settings` returns data of the form:
+      ```json
+      {
+        "settings": {
+          "index": {
+            "number_of_shards": 18,
+            "number_of_replicas": 0,
+            "refresh_interval": "-1"
+          },
+          "analysis": {...},
+        },
+        "mapping": {...}
+      }
+      ```
+      However, the `index.get` function returns data of the form:
+      ```json
+      {
+        "<index-name>": {
+          "settings": {
+            "index": {
+              "number_of_shards": "18",
+              "number_of_replicas": 0,
+              "refresh_interval": "-1",
+              ...
+              "analysis": {
+                ...
+              }
+            }
+          },
+          "mappings": {...}
+        }
+      }
+      ```
+      The returned information will need to be extracted and converted into the
+      form returned by `es_mapping::index_settings` before it is merged with the
+      new configuration.
```

Comment:
Just a nit, but the focus on the API response being different from `es_mapping::index_settings` rather than the expected input of the Elasticsearch create index API feels strange and potentially misleading as to the reason for it. It feels like just a holdover from when this IP originally used `es_mapping::index_settings` instead of requesting the existing settings from the API that distracts from the actual goal of sending those settings to Elasticsearch. `es_mapping::index_settings`'s return format is only relevant insofar as it currently happens to return a dict exactly compatible with ES's create index API. That's not a necessary relationship, however, so it feels like a distraction from what goal the transformation of the index settings API response needs to be (to match ES create index input API).

Linking to the ES create index API request body documentation would be more direct and clearer, especially as that `es_mapping::index_settings` function is pretty much irrelevant for any of the other stuff this DAG does.

https://www.elastic.co/guide/en/elasticsearch/reference/7.17/indices-create-index.html#indices-create-api-request-body